pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentOpenLongPairRangeSet.java

```diff
@@ -0,0 +1,474 @@
+                return;
+            }
+            isEmpty.set(val.isEmpty());
+        });
+        return isEmpty.get();
+    }
+
+    @Override
+    public void clear() {
+        rangeBitSetMap.clear();
+        updatedAfterCached = true;
+    }
+
+    @Override
+    public Range<T> span() {
+        Entry<Long, BitSet> firstSet = rangeBitSetMap.firstEntry();
+        Entry<Long, BitSet> lastSet = rangeBitSetMap.lastEntry();
+        int first = firstSet.getValue().nextSetBit(0);
+        int last = lastSet.getValue().previousSetBit(lastSet.getValue().size());
+        return Range.openClosed(consumer.apply(firstSet.getKey(), first - 1), consumer.apply(lastSet.getKey(), last));
+    }
+
+    @Override
+    public List<Range<T>> asRanges() {
+        List<Range<T>> ranges = new ArrayList<>();
+        forEach((range) -> {
+            ranges.add(range);
+            return true;
+        });
+        return ranges;
+    }
+
+    @Override
+    public void forEach(RangeProcessor<T> action) {
+        forEach(action, consumer);
+    }
+
+    @Override
+    public void forEach(RangeProcessor<T> action, LongPairConsumer<? extends T> consumer) {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        rangeBitSetMap.forEach((key, set) -> {
+            if (completed.get()) {
+                return;
+            }
+            if (set.isEmpty()) {
+                return;
+            }
+            int first = set.nextSetBit(0);
+            int last = set.previousSetBit(set.size());
+            int currentClosedMark = first;
+            while (currentClosedMark != -1 && currentClosedMark <= last) {
+                int nextOpenMark = set.nextClearBit(currentClosedMark);
+                Range<T> range = Range.openClosed(consumer.apply(key, currentClosedMark - 1),
+                        consumer.apply(key, nextOpenMark - 1));
+                if (!action.process(range)) {
+                    completed.set(true);
+                    break;
+                }
+                currentClosedMark = set.nextSetBit(nextOpenMark);
+            }
+        });
+    }
+
+    @Override
+    public Range<T> firstRange() {
+        Entry<Long, BitSet> firstSet = rangeBitSetMap.firstEntry();
+        int lower = firstSet.getValue().nextSetBit(0);
+        int upper = Math.max(lower, firstSet.getValue().nextClearBit(lower) - 1);
+        return Range.openClosed(consumer.apply(firstSet.getKey(), lower - 1), consumer.apply(firstSet.getKey(), upper));
+    }
+
+    @Override
+    public int size() {
+        if (updatedAfterCached) {
+            AtomicInteger size = new AtomicInteger(0);
+            forEach((range) -> {
+                size.getAndIncrement();
+                return true;
+            });
+            cachedSize = size.get();
+            updatedAfterCached = false;
+        }
+        return cachedSize;
+    }
+
+    @Override
+    public String toString() {
+        if (updatedAfterCached) {
```

Comment:
The `toString()` should typically never be called in "real-world"  (other than tests or debug logs). Updating the `volatile` flag will have a small overhead that we could avoid by removing this optimization.

Reply:
actually `toString()` is also called everytime when we call stats or stats-internal. so, this flag can help us to use cached output if it's not changed and can save cpu.